### PR TITLE
feat: add custom assignment rules evaluation

### PR DESCRIPTION
**Pull request by joalves**
## Summary
- Add `evaluateRules()` to `AudienceMatcher` supporting the new `rules[].or[]` payload structure with environment scoping and JsonExpr conditions
- Integrate rules evaluation into `Context._assign()` — rules are checked after audience filter but before `audienceStrict` / traffic split
- Add `getEnvironment()` to `Client` so the Context can access the environment name for rule filtering
- Add `targetingRule` flag to exposure events for explicit identification in analytics
- Validate `rule.variant` is a number; skip invalid rules instead of aborting iteration

## Assignment Flags

Rule-matched assignments set `assigned=true, eligible=true, overridden=true, targetingRule=true`.

### All possible assignment states

| State | `assigned` | `eligible` | `overridden` | `fullOn` | `custom` | `audienceMismatch` | `targetingRule` |
|---|---|---|---|---|---|---|---|
| **Not running** | false | true | false | false | false | false | false |
| **Override** | false | true | true | false | false | false | false |
| **Audience strict mismatch** | false | true | false | false | false | true | false |
| **Rule match** | true | true | true | false | false | * | true |
| **Normal (eligible)** | true | true | false | false | false | false | false |
| **Normal (not eligible)** | true | false | false | false | false | false | false |
| **Custom assignment** | true | true | false | false | true | false | false |
| **Full-on** | true | true | false | true | false | false | false |

`*` = `audienceMismatch` is set independently before rules, can be either value.

### Engine bitmask

`targetingRule` uses bit 8 (value 256). The existing filter `bitAnd(flags, 207) = 3` has bit 8 as `0` in the mask, so it's automatically ignored — no engine query changes needed. The engine's `Exposure.java` and `EventForwardingManager.java` will need the new field when implementing server-side support.

## Test Plan
- [x] 21 unit tests for `evaluateRules` in matcher
- [x] 16 integration tests in context (flags, cache invalidation, overrides, variables)
- [x] All 299 tests pass (29 suites)

## Summary by CodeRabbit

* **New Features**
  * Environment-scoped assignment rules can force variant selection; exposures and published payloads now include a ruleOverride flag and variables honour rule-selected variants.

* **Behaviour**
  * Rules use first-match semantics and environment gating; malformed/invalid or out-of-range rule results fall back to normal assignment; explicit overrides take precedence; caching and assignment recompute when attributes or rule context change.

* **Tests**
  * Expanded coverage for rule evaluation, environment gating, precedence, cache invalidation and many edge cases.

### CODE REVIEW

File: src/matcher.ts [hunk 1]
```diff
@@ -17,5 +17,40 @@ export class AudienceMatcher {
...
+						const conditions = rule.and;
+						if (!conditions || (Array.isArray(conditions) && conditions.length === 0)) {
+							return rule.variant;
+						}
+						if (Array.isArray(conditions)) {
+							const result = this._jsonExpr.evaluateBooleanExpr({ and: conditions }, vars);
+							if (result === true) {
+								return rule.variant;
+							}
+						}
```
[the remark this answers is not in the document]
Comment (joalves) [reply src/matcher.ts]: Fixed in commit 003dada. Both return sites now validate `typeof rule.variant === "number"` before returning. Added tests for missing variant and non-number variant.

File: src/context.ts [hunk 2]
```diff
@@ -536,7 +538,20 @@ export default class Context {
...
 
-				if (experiment.data.audienceStrict && assignment.audienceMismatch) {
+				const ruleVariant =
+					experiment.data.audience && experiment.data.audience.length > 0
+						? this._audienceMatcher.evaluateRules(
+								experiment.data.audience,
+								this._environmentName,
+								this._getAttributesMap()
+						  )
+						: null;
```
[the remark this answers is not in the document]
Comment (joalves) [reply src/context.ts]: Good catch. Fixed — the `audienceMatches` cache check now also re-evaluates rules when attributes change. The computed `ruleVariant` is stored on the assignment and compared on subsequent calls. If the rule result differs from the cached one, the assignment is invalidated and recomputed.

Added two tests:
- Rule starts not-matching → attribute change → rule matches → returns rule variant (not stale cached normal assignment)
- Rule starts matching → attribute change → rule stops matching → returns normal assignment (not stale cached rule variant)

File: src/context.ts [hunk 3]
```diff
@@ -536,7 +538,20 @@ export default class Context {
...
+
+				if (ruleVariant !== null) {
+					assignment.assigned = true;
+					assignment.variant = ruleVariant;
+					assignment.overridden = true;
```
[the remark this answers is not in the document]
Comment (joalves) [reply src/context.ts]: We intentionally use `overridden=true` for rule matches because the engine's bitmask filter (`bitAnd(flags, 207) = 3`) uses the `overridden` bit to exclude non-random assignments from statistical analysis. Using a separate flag would require engine changes.

The theoretical concern about `override("exp", 1)` reusing a cached rule assignment (with `assigned=true` instead of `assigned=false`) only occurs when the override variant equals the rule variant. In that case, the user gets the same variant either way — the only difference is the `assigned` flag in the exposure, which has no practical impact since both are excluded from analysis via `overridden=true`.

Adding a `source` field would increase complexity across all SDKs without meaningful benefit. Not addressing this edge case.

[the remark this answers is not in the document]
Comment (joalves) [reply src/context.ts]: On second look, the edge case is real: when an override uses the same variant as a rule-cached assignment, the cache would return stale flags (`assigned=true` instead of `assigned=false`).

Fixed by adding `!assignment.assigned` to the override cache check. Real overrides always have `assigned=false`, rule-cached assignments have `assigned=true`, so the cache correctly misses and recomputes with proper override flags.

No need for a `source` field — the existing flags are sufficient to distinguish the two cases.

**Comment by joalves** [comment]
### Code review

Found 2 issues (both now fixed in f32df8e):

1. `return null` on invalid variant aborts all rule iteration -- a matched rule with a non-numeric `variant` would `return null`, preventing subsequent valid rules from being evaluated. Changed to `continue` so the loop skips invalid rules and tries the next one.

https://github.com/absmartly/javascript-sdk/blob/3b9ab8cee25f05000de56000ddbb1a51fbc61892/src/matcher.ts#L33-L41

2. `audienceMatches` does not persist `ruleVariant` back to the cached assignment -- after re-evaluating rules and confirming the variant hasn't changed, `assignment.attrsSeq` was updated but `assignment.ruleVariant` was not. This breaks the invariant that if `attrsSeq` is up-to-date then `ruleVariant` is up-to-date, causing unnecessary re-evaluations on subsequent attribute changes.

https://github.com/absmartly/javascript-sdk/blob/3b9ab8cee25f05000de56000ddbb1a51fbc61892/src/context.ts#L476-L481

File: src/context.ts [hunk 4]
```diff
@@ -60,6 +60,7 @@ type Assignment = {
 	fullOn: boolean;
 	custom: boolean;
 	audienceMismatch: boolean;
+	ruleVariant?: number | null;
```
Comment (calthejuggler) [inline src/context.ts:66]: It's effectively just an override that's configured from the web console, right? 🤔 Why do we need this?

File: src/context.ts [hunk 5]
```diff
@@ -528,15 +538,30 @@ export default class Context {
...
+					ruleVariant = this._audienceMatcher.evaluateRules(experiment.data.audience, this._environmentName, attrs);
 				}
 
-				if (experiment.data.audienceStrict && assignment.audienceMismatch) {
+				assignment.ruleVariant = ruleVariant;
+
+				if (ruleVariant !== null) {
+					// Rule-matched: assigned=true + overridden=true
+					// SDK overrides: assigned=false + overridden=true
+					// This distinction lets analytics differentiate the two cases
```
Comment (calthejuggler) [inline src/context.ts]: Does the engine definitely support this? 🤔 Feels like quite a lot of work, for something that could be simpler. I guess if we wanted analytics to be able to tell the difference, we could make it the same as an override, but the ruleVariant is set? Or we could change `ruleVariant` to be a `ruleOverride: boolean` - then we get the best of both worlds in the future, but we can keep the assigned/overridden logic the same.

File: src/matcher.ts [hunk 6]
```diff
@@ -17,5 +17,37 @@ export class AudienceMatcher {
...
+					}
+				}
+			}
+		} catch (error) {
+			console.error(error);
+		}
+		return null;
+	}
```
Comment (calthejuggler) [inline src/matcher.ts]: We can reduce nesting here with some early returns 🙂 Something like:
```suggestion
	evaluateRules(audienceString: string, environmentName: string | null, vars: Record<string, unknown>): number | null {
	    let audience
		try {
			audience = JSON.parse(audienceString);
		} catch (error) {
			console.error(error);
			return null;
		}
		
		if (!audience || !Array.isArray(audience.rules)) return null;

		for (const ruleGroup of audience.rules) {
			if (!ruleGroup || !Array.isArray(ruleGroup.or)) continue;
			for (const rule of ruleGroup.or) {
				if (Array.isArray(rule.environments) && rule.environments.length > 0) {
					if (environmentName == null || !rule.environments.includes(environmentName)) {
						continue;
					}
				}
				
				if (typeof rule.variant !== "number") continue;
				
				const conditions = rule.and;
				
				if (!conditions || (Array.isArray(conditions) && conditions.length === 0)) {
					return rule.variant;
				}
				
				if (!Array.isArray(conditions)) continue;

				const result = this._jsonExpr.evaluateBooleanExpr({ and: conditions }, vars);
				if (result === true) {
					return rule.variant;
				}
			}
		}
		
		return null;
	}
```

File: src/__tests__/context.test.js [hunk 7]
```diff
@@ -2025,6 +2025,364 @@ describe("Context", () => {
...
+									or: [
+										{
+											name: "Production Only",
+											and: [{ eq: [{ var: "country" }, { value: "US" }] }],
+											environments: ["production"],
+											variant: 1,
+										},
+									],
+								},
+							],
```
Comment (calthejuggler) [inline src/__tests__/context.test.js]: We should add some tests for multiple `or`s and multiple `and`s (and maybe multiple environments?)

File: src/__tests__/context.test.js [hunk 8]
```diff
@@ -2025,6 +2025,364 @@ describe("Context", () => {
 		});
 	});
 
+	describe("rules evaluation", () => {
+		// Uses exp_test_abc (3 variants, normal assignment = 2) with rules forcing variant 1
+		// This ensures tests are meaningful: rule variant (1) differs from normal assignment (2)
```
Comment (calthejuggler) [inline src/__tests__/context.test.js]:
```suggestion
```

File: src/__tests__/context.test.js [hunk 9]
```diff
@@ -2025,6 +2025,364 @@ describe("Context", () => {
...
+												environments: [],
+												variant: 1,
+											},
+											{
+												name: "GB Users",
+												and: [{ eq: [{ var: "country" }, { value: "GB" }] }],
+												environments: [],
+												variant: 2,
+											},
+										],
```
Comment (calthejuggler) [inline src/__tests__/context.test.js]: Ah there's a double `or` here - maybe we can add some extra tests for these cases. Plus some multiple `and`s and multiple `environments` 🙂

**Review by calthejuggler — COMMENTED** [review]
Looking really good so far 🙂 
Good to put it in a separate PR, but let's not forget to add the SDK version to the payload before we publish this.

File: src/context.ts (hunk 5 above)
Comment (joalves) [reply src/context.ts]: Renamed to `ruleOverride` as you suggested. The engine uses a UInt32 for flags with only 8 of 32 bits used, so bit 8 is free — no schema changes needed. The flag is set alongside `overridden=true` so the existing `bitAnd(flags, 207) = 3` filter still correctly excludes rule-forced assignments from statistical analysis.

File: src/matcher.ts (hunk 6 above)
Comment (joalves) [reply src/matcher.ts]: Applied — refactored to early returns as you suggested. Moved the try/catch to only wrap JSON.parse, then early-return on invalid audience. Also added `if (!Array.isArray(conditions)) continue` for the last guard.

File: src/__tests__/context.test.js (hunk 7 above)
Comment (joalves) [reply src/__tests__/context.test.js]: Added tests for:
- Multiple `and` conditions (country=US AND user_type=internal)
- Multiple environments (`["production", "staging"]`)
- Multiple `or` rules (US→1, GB→2, FR→0, DE→normal assignment)

File: src/__tests__/context.test.js (hunk 8 above)
Comment (joalves) [reply src/__tests__/context.test.js]: Removed.

File: src/__tests__/context.test.js (hunk 9 above)
Comment (joalves) [reply src/__tests__/context.test.js]: Added context integration tests covering multiple `or` rules, multiple `and` conditions, and multiple environments. The matcher unit tests already had coverage for multiple rule groups and multiple rules within an `or` block.

File: src/context.ts (hunk 4 above)
Comment (joalves) [reply src/context.ts:66]: This is the cached rule evaluation result stored on the assignment — it's used by the `audienceMatches` cache check to detect when a rule's matching variant changes due to attribute updates. Without it, changing an attribute used only by a rule (not the audience filter) wouldn't invalidate the cached assignment. It's internal to the SDK, not sent in the exposure event. Renamed it to `ruleVariant` to be clearer — it stores which variant the rule evaluated to (or null if no rule matched).

Comment (joalves) [reply src/context.ts:66]: To clarify further: we need this field because the existing cache checks don't cover rule changes. `experimentMatches` compares id, unitType, iteration, fullOnVariant, and trafficSplit — but not the variant itself. `audienceMatches` only checks `audienceMismatch`. So if an attribute changes that affects a rule condition but not the audience filter, neither check would detect it — the cache would return a stale assignment. `ruleVariant` fills that gap: when `attrsSeq` advances, we re-evaluate rules and compare against the stored `ruleVariant`. If it changed, the cache invalidates and recomputes. Without it, we'd have to recompute on every attribute change regardless.

File: src/context.ts [hunk 10]
```diff
@@ -482,7 +494,7 @@ export default class Context {
 		if (experimentName in this._assignments) {
 			const assignment = this._assignments[experimentName];
 			if (hasOverride) {
-				if (assignment.overridden && assignment.variant === this._overrides[experimentName]) {
+				if (assignment.overridden && !assignment.assigned && assignment.variant === this._overrides[experimentName]) {
```
Comment (calthejuggler) [inline src/context.ts]: This isn't necessary anymore - right?

File: src/context.ts [hunk 11]
```diff
@@ -528,15 +541,28 @@ export default class Context {
...
 					}
+
+					ruleVariant = this._audienceMatcher.evaluateRules(experiment.data.audience, this._environmentName, attrs);
 				}
 
-				if (experiment.data.audienceStrict && assignment.audienceMismatch) {
+				assignment.ruleVariant = ruleVariant;
+
+				if (ruleVariant !== null) {
+					assignment.assigned = true;
```
Comment (calthejuggler) [inline src/context.ts]: Still not sure if `assigned: true` here is a true statement 🤔

File: src/context.ts (hunk 10 above)
Comment (joalves) [reply src/context.ts]: Correct — removed in 126cc67. Rules no longer set `overridden=true`, so there's no cache collision with SDK overrides anymore.

File: src/context.ts (hunk 11 above)
Comment (joalves) [reply src/context.ts]: Agreed — changed to `assigned: false` in 126cc67. Rule-forced users are not real experiment participants, so `assigned=false` is correct. The `ruleOverride=true` flag is now the sole identifier for rule-matched assignments.

File: src/context.ts [hunk 12]
```diff
@@ -528,15 +541,25 @@ export default class Context {
...
+
+					ruleVariant = this._audienceMatcher.evaluateRules(experiment.data.audience, this._environmentName, attrs);
 				}
 
-				if (experiment.data.audienceStrict && assignment.audienceMismatch) {
+				assignment.ruleVariant = ruleVariant;
+
+				if (ruleVariant !== null) {
+					assignment.variant = ruleVariant;
+					assignment.ruleOverride = true;
```
Comment (calthejuggler) [inline src/context.ts:597]: This isn't quite what I meant 🤔 We want the exclusion rules to basically be the same as an override (and present as such to the engine). To me that would mean setting `assignment.overriden` here, and simply using `ruleOverride` for metadata later.
```suggestion
					assignment.variant = ruleVariant;
					assignment.overridden = true;
					assignment.ruleOverride = true;
```
This way - the engine can function exactly how it does today, but we have the added benefit of being able to work out whether the override was from a direct code override, or a rule in the UI

File: src/context.ts [hunk 13]
```diff
@@ -729,7 +753,7 @@ export default class Context {
 					this._queueExposure(experimentName, assignment);
 				}
 
-				if (key in assignment.variables && (assignment.assigned || assignment.overridden)) {
+				if (key in assignment.variables && (assignment.assigned || assignment.overridden || assignment.ruleOverride)) {
```
Comment (calthejuggler) [inline src/context.ts:792]: We then wouldn't need to add these checks, and the tests should pass just the same

File: src/context.ts (hunk 12 above)
Comment (joalves) [reply src/context.ts:597]: We originally set `overridden=true` to pass the engine's `bitAnd(flags, 207) = 3` filter, but that was abusing the flag just to get excluded from analysis. Now that we set `assigned=false`, the filter already excludes rule-matched assignments (it requires `assigned=true`), so we don't need to piggyback on `overridden` anymore.

This gives us proper semantics:

| State | `assigned` | `overridden` | `ruleOverride` |
|---|---|---|---|
| **Rule match** | false | false | true |
| **SDK override** | false | true | false |
| **Normal assignment** | true | false | false |

- `overridden=true` means "SDK developer override" (its original meaning)
- `ruleOverride=true` means "rule-forced from the web console"
- Both are excluded from analysis, but for different reasons (`overridden` via the bitmask, `ruleOverride` via `assigned=false`)

It also avoids the cache issue we had earlier — when rules set `overridden=true`, the override cache check (`assignment.overridden && assignment.variant === this._overrides[...]`) would incorrectly match rule-cached assignments. We had to add a `!assignment.assigned` workaround. With `overridden=false` on rules, that problem doesn't exist and the code stays clean.

For the `variableValue` check — `overridden` already passes that condition for SDK overrides, and `ruleOverride` passes it for rules. Each flag handles its own case without conflation.

File: src/context.ts (hunk 13 above)
Comment (joalves) [reply src/context.ts:792]: True, we wouldn't need these checks here, but we'd have to add the `!assignment.assigned` workaround back to the override cache check instead, to prevent rule-cached assignments from colliding with SDK overrides. So we're trading one check for another, and losing the clean semantics on the way.

Comment (calthejuggler) [reply src/context.ts:792]: Would we? 🤔 It would be checked already by the `overridden` flag.

Comment (joalves) [reply src/context.ts:792]: The `overridden` flag alone doesn't cover the edge case where the override variant **equals** the rule variant.

Consider this sequence (assuming rules set `overridden: true` as you suggested):

1. `treatment("exp")` → rule matches → cached as `{overridden: true, assigned: true, variant: 1}`
2. `override("exp", 1)` → override cache check: `assignment.overridden && assignment.variant === this._overrides["exp"]` → `true && 1 === 1` → **cache hit** → returns the stale rule assignment

The problem: it returns the rule-cached assignment with `assigned: true`, but a real override should have `assigned: false`. The `overridden` flag is `true` in both cases, so it can't tell them apart.

The `!assignment.assigned` guard broke that false cache hit: rules have `assigned: true`, real overrides have `assigned: false`, so the check becomes `assignment.overridden && !assignment.assigned && variant === override` — which correctly misses on rule-cached assignments and forces a recompute.

That said, this is now moot — in the current code rules don't set `overridden: true` anymore, so there's no collision. The `ruleOverride` flag is the sole identifier.

**Review by calthejuggler — APPROVED** [review]
